ldelf/ta_elf.c

```diff
@@ -0,0 +1,811 @@
+	    ehdr->e_shentsize != sizeof(Elf64_Shdr))
+		return TEE_ERROR_BAD_FORMAT;
+
+
+	elf->is_32bit = false;
+	elf->e_entry = ehdr->e_entry;
+	elf->e_phoff = ehdr->e_phoff;
+	elf->e_shoff = ehdr->e_shoff;
+	elf->e_phnum = ehdr->e_phnum;
+	elf->e_shnum = ehdr->e_shnum;
+	elf->e_phentsize = ehdr->e_phentsize;
+	elf->e_shentsize = ehdr->e_shentsize;
+
+	return TEE_SUCCESS;
+}
+#else /*ARM64*/
+static TEE_Result e64_parse_ehdr(struct ta_elf *elf __unused,
+				 Elf64_Ehdr *ehdr __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+#endif /*ARM64*/
+
+static void read_dyn(struct ta_elf *elf, vaddr_t addr,
+		     size_t idx, unsigned int *tag, size_t *val)
+{
+	if (elf->is_32bit) {
+		Elf32_Dyn *dyn = (Elf32_Dyn *)(addr + elf->load_addr);
+
+		*tag = dyn[idx].d_tag;
+		*val = dyn[idx].d_un.d_val;
+	} else {
+		Elf64_Dyn *dyn = (Elf64_Dyn *)(addr + elf->load_addr);
+
+		*tag = dyn[idx].d_tag;
+		*val = dyn[idx].d_un.d_val;
+	}
+}
+
+static void e32_save_symtab(struct ta_elf *elf, size_t tab_idx)
+{
+	Elf32_Shdr *shdr = elf->shdr;
+	size_t str_idx = shdr[tab_idx].sh_link;
+
+	elf->dynsymtab = (void *)(shdr[tab_idx].sh_addr + elf->load_addr);
+	assert(!(shdr[tab_idx].sh_size % sizeof(Elf32_Sym)));
+	elf->num_dynsyms = shdr[tab_idx].sh_size / sizeof(Elf32_Sym);
+
+	elf->dynstr = (void *)(shdr[str_idx].sh_addr + elf->load_addr);
+	elf->dynstr_size = shdr[str_idx].sh_size;
+}
+
+static void e64_save_symtab(struct ta_elf *elf, size_t tab_idx)
+{
+	Elf64_Shdr *shdr = elf->shdr;
+	size_t str_idx = shdr[tab_idx].sh_link;
+
+	elf->dynsymtab = (void *)(vaddr_t)(shdr[tab_idx].sh_addr +
+					   elf->load_addr);
+	assert(!(shdr[tab_idx].sh_size % sizeof(Elf64_Sym)));
+	elf->num_dynsyms = shdr[tab_idx].sh_size / sizeof(Elf64_Sym);
+
+	elf->dynstr = (void *)(vaddr_t)(shdr[str_idx].sh_addr + elf->load_addr);
+	elf->dynstr_size = shdr[str_idx].sh_size;
+}
+
+static void save_symtab(struct ta_elf *elf)
+{
+	size_t n = 0;
+
+	if (elf->is_32bit) {
+		Elf32_Shdr *shdr = elf->shdr;
+
+		for (n = 0; n < elf->e_shnum; n++) {
+			if (shdr[n].sh_type == SHT_DYNSYM) {
+				e32_save_symtab(elf, n);
+				break;
+			}
+		}
+	} else {
+		Elf64_Shdr *shdr = elf->shdr;
+
+		for (n = 0; n < elf->e_shnum; n++) {
+			if (shdr[n].sh_type == SHT_DYNSYM) {
+				e64_save_symtab(elf, n);
+				break;
+			}
+		}
+
+	}
+}
+
+static void init_elf(struct ta_elf *elf)
+{
+	TEE_Result res = TEE_SUCCESS;
+	vaddr_t va = 0;
+	uint32_t flags = PTA_SYSTEM_MAP_FLAG_SHAREABLE;
+	const size_t max_align = 0x10000;
+
+	res = sys_open_ta_bin(&elf->uuid, &elf->handle);
+	if (res)
+		err(res, "sys_open_ta_bin(%pUl)", (void *)&elf->uuid);
+
+	/*
+	 * Map it read-only executable when we're loading a library where
+	 * the ELF header is included in a load segment.
+	 */
+	if (!elf->is_main)
+		flags |= PTA_SYSTEM_MAP_FLAG_EXECUTABLE;
+	/*
+	 * Add 1Mb pad at end in case a library with this large alignment
+	 * has been mapped before. We want to avoid ending up in a hole in
+	 * the mapping of a library.
```

Comment:
Why? Many libraries are likely to be less than 1 MB and could fit in that hole.

Reply:
Because we don't have the patch "core: pta_system: support memory remap" at this stage so we have to stick with the address we get here.